CoreAlgorithmPlugins/src/au/gov/asd/tac/constellation/plugins/algorithms/sna/centrality/PagerankCentralityPlugin.java

```diff
@@ -95,118 +92,181 @@ public PluginParameters createParameters() {
-                                || (vertexId < neighbourId && edgeDirection == GraphConstants.DOWNHILL)
-                                || (vertexId > neighbourId && edgeDirection == GraphConstants.UPHILL)) {
-                            int neighbourOutCount = 0;
-                            final int transactionCount = graph.getVertexTransactionCount(neighbourId);
-                            for (int transactionPosition = 0; transactionPosition < transactionCount; transactionPosition++) {
-                                final int transactionId = graph.getVertexTransaction(neighbourId, transactionPosition);
-                                final int transactionDirection = graph.getTransactionDirection(transactionId);
-                                final int transactionSourceId = graph.getTransactionSourceVertex(transactionId);
-                                final int transactionDestinationId = graph.getTransactionDestinationVertex(transactionId);
-                                if (transactionSourceId != transactionDestinationId) {
-                                    final int otherId = neighbourId == transactionSourceId ? transactionDestinationId : transactionSourceId;
-                                    if ((treatUndirectedBidirectional && transactionDirection == GraphConstants.FLAT)
-                                            || (neighbourId < otherId && transactionDirection == GraphConstants.UPHILL)
-                                            || (neighbourId > otherId && transactionDirection == GraphConstants.DOWNHILL)) {
-                                        neighbourOutCount++;
-                                    }
-                                }
-                            }
+        }
+
+        // update the graph with pagerank values
+        PagerankVertex.commitPageranksToGraph();
+    }
+    
+    private static class PagerankVertex {
+        private static GraphWriteMethods graph;
+        private static final Set<PagerankVertex> sinks = new HashSet<>();
+        private static boolean treatUndirectedBidirectional;
+        private static boolean normaliseByAvailable;
+        private static double dampingFactor = 0;
+        private static int vertexCount = 0;
+        private static PagerankVertex[] pagerankVertices;
+        private static double sinkPagerankContribution = 0;
+        private static double delta = 0;
+        private static int pagerankAttribute;
+        private static double baseContribution;
+        private final int vertexId;
+        private final Set<PagerankVertex> neighbours = new HashSet<>();
+        private int outCount = 0;
+        private double pagerank;
+        private double pagerankContribution;
+        private double stagedPagerank;
+        private boolean isSink = false;
 
-                            vertexInLinks.get(vertexId).add(Tuple.create(neighbourId, neighbourOutCount));
+        private PagerankVertex(final int vertexPosition) {
```

Comment:
It feels weird to me that this class both represents a Pagerank controller for a graph as well as an individual Pagerank result for a vertex. Maybe you have a good reason for doing this, but it feels like it could use some logical seperation.

Reply:
Nope no good reason, just hadn't thought about it.
I think I agree, i'm going to make this change and see how it feels.